sections/advanced/reaching-ui-state.md

```diff
@@ -0,0 +1,222 @@
+
+```javascript
+function stubRecorder(pathToJson) {
+  const xhrData = []; // an empty array to hold the data
+  cy.server({ // if recording, save the response data in the array
+    onResponse: (response) => {
+      const url = response.url;
+      const method = response.method;
+      const data = response.response.body;
+      // We push a new entry into the xhrData array
+      xhrData.push({ url, method, data });
+    }
+  });
+  // this is used as a ‘filter’ for the data you want to record.
+  // You can specify the methods and routes that are recorded
+  cy.log('recording!');
+  cy.route({
+    method: 'GET',
+    url: '*',
+  });
+  // if recording, after the test runs, create a fixture file with the recorded data
+  after(function () {
+    cy.writeFile(`./cypress/fixtures/${pathToJson}.json`, xhrData);
+    cy.log(`Wrote ${xhrData.length} XHR responses to local file ${pathToJson}.json`);
+  });
+}
+
+/** Plays recorded fixture with all required network data as json*/
+function playStubbedFixture(stateFixture) {
+  cy.log(`playing fixture from ${stateFixture}`);
+  cy.fixture(stateFixture, { timeout: 15000 }) // the fixture file may be large and take time in CI
+    .then(data => {
+      for (let i = 0, length = data.length; i < length; i++) {
+        cy.route(data[i].method, data[i].url, data[i].data);
+      }
+    }).as(`stateFixture_stub`);
+}
+
+/** Visits the stubbed state */
+function visitStubbedState(stubFile, url, wait: boolean = true) {
+  playStubbedFixture(stubFile);
+  cy.visit(url);
+  if (wait) { // sometimes you do not want to wait for network, this gives you the option
+    cy.wait('@stateFixture_stub', { timeout: 15000 });
+  }
+}
+
+//////////
+// usage
+
+// recording network
+it('should run your test', function () {
+  stubrecorder('jsonfileNameForNetworkData');
+
+  // your original test
+
+  cy.wait(5000); // one-time wait so that the after() step records all the network without missing anything
+
+  // the rest of your original test
+});
+
+// playing the stubbed network
+it('should run your test', function () {
+  // every time we visit this endpoint, all network will be stubbed
+  // double check this by observing (XHR stubbed) network responses in the test runner
+  visitStubbedState('jsonfileNameForNetworkData', '/endpoint');
+
+  // the rest of your original test
+});
+```
+
+
+<br/><br/>
+
+### Application actions
+
+Cypress gives complete control of your application. You can bypass the page object abstraction layer (which is detached from your application) access the UI directly via `cy.get()` , have access to the API and database, and even access the source code.
+
+Component testing is an experimental feature in Cypress as of version 4.7.0, albeit some of the functionality is available. Refer to component testing articles under tools section for experiments with React and Storybook on this topic.
+
+![devtools > network tab](../../assets/images/ui-state/app-action.PNG)
```

Comment:
I think that this image could be removed because the user has not enough context about the Page Object pattern here

Reply:
I wish that were true. Page Object pattern too well known that it is hard to convince people that it is outdated and there is a better way.

But, again, this is all my perspective. If you think it confuses the reader, please remove it.

Reply:
I think that this part

> (which is detached from you application)

is self-explanatory, if you want, add a link to something deeper directly inside the parenthesis. I don't like the image because it only reports the four layers but an image is not needed to express the layers themselves (and they are a side-topic compared to the topic of this chapter)

p.s. I noticed just now the *(which is detached from **you** application)* typo 😁

Reply:
removed the image